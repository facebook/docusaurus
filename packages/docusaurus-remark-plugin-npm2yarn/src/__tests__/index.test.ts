/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import path from 'path';
import vfile from 'to-vfile';
import dedent from 'dedent';
import npm2yarn from '../index';

const process = async (
  content: any,
  options?: Parameters<typeof npm2yarn>[0],
) => {
  const {remark} = await import('remark');
  const {default: mdx} = await import('remark-mdx');

  const result = await remark()
    .use(mdx)
    .use(npm2yarn, options)
    .process(content);

  return result.value;
};

const processFixture = async (
  name: string,
  options?: Parameters<typeof npm2yarn>[0],
) => {
  const filePath = path.join(__dirname, '__fixtures__', `${name}.md`);
  const file = await vfile.read(filePath);
  return process(file, options);
};

describe('npm2yarn plugin', () => {
  it('works with simplest md', async () => {
    const result = await process(dedent`
    # Title

    Hey

    \`\`\`bash npm2yarn
    npm install test
    \`\`\`

    `);

    expect(result).toMatchSnapshot();
  });

  it('works on installation file', async () => {
    const result = await processFixture('installation');

    expect(result).toMatchSnapshot();
  });

  it('works on plugin file', async () => {
    const result = await processFixture('plugin');

    expect(result).toMatchSnapshot();
  });

  it('works with common commands', async () => {
    const result = await processFixture('conversion-test', {sync: true});

    expect(result).toMatchSnapshot();
  });

  it('works with sync option', async () => {
    const result = await processFixture('plugin', {sync: true});

    expect(result).toMatchSnapshot();
  });

  it('does not work when language is not set', async () => {
    const result = await processFixture('syntax-not-properly-set');

    expect(result).toMatchSnapshot();
  });

  it('does not re-import tabs components when already imported above', async () => {
    const result = await processFixture('import-tabs-above');

    expect(result).toMatchSnapshot();
  });

  it('does not re-import tabs components when already imported below', async () => {
    const result = await processFixture('import-tabs-below');

    expect(result).toMatchSnapshot();
  });

  it('does not re-import tabs components real-world multiple npm2yarn usage', async () => {
    const result = await processFixture('multiple');
    expect(result).toMatchSnapshot();
  });

  it('work with yarn converter', async () => {
    const result = await processFixture('plugin', {converters: ['yarn']});

    expect(result).toMatchSnapshot();
  });

  it('work with pnpm converter', async () => {
    const result = await processFixture('plugin', {converters: ['pnpm']});

    expect(result).toMatchSnapshot();
  });

  it('work with custom converter', async () => {
    const result = await processFixture('plugin', {
      converters: [['Turbo', (code) => code.replace(/npm/g, 'turbo')]],
    });

    expect(result).toMatchSnapshot();
  });
});
