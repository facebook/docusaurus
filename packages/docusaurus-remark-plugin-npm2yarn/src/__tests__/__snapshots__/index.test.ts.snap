// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`npm2yarn plugin does not re-import tabs components real-world multiple npm2yarn usage 1`] = `
"import Tabs from '@theme/Tabs'
import TabItem from '@theme/TabItem'

# Title

<Tabs>
  <TabItem value="npm">
    \`\`\`bash
      $ npm install --global docusaurus
    \`\`\`
  </TabItem>

  <TabItem value="yarn" label="Yarn">
    \`\`\`bash
      $ yarn global add docusaurus
    \`\`\`
  </TabItem>

  <TabItem value="pnpm" label="pnpm">
    \`\`\`bash
      $ pnpm add --global docusaurus
    \`\`\`
  </TabItem>
</Tabs>

<div className="nested-npm2yarn">
  <Tabs>
    <TabItem value="npm">
      \`\`\`bash
      npm install
      \`\`\`
    </TabItem>

    <TabItem value="yarn" label="Yarn">
      \`\`\`bash
      yarn install
      \`\`\`
    </TabItem>

    <TabItem value="pnpm" label="pnpm">
      \`\`\`bash
      pnpm install
      \`\`\`
    </TabItem>
  </Tabs>
</div>

\`\`\`bash
echo "no npm2yarn here"
\`\`\`

## Subtitle

<Tabs>
  <TabItem value="npm">
    \`\`\`bash
    yarn add @docusaurus/core
    \`\`\`
  </TabItem>

  <TabItem value="yarn" label="Yarn">
    \`\`\`bash
    yarn add @docusaurus/core
    \`\`\`
  </TabItem>

  <TabItem value="pnpm" label="pnpm">
    \`\`\`bash
    yarn add @docusaurus/core
    \`\`\`
  </TabItem>
</Tabs>
"
`;

exports[`npm2yarn plugin does not re-import tabs components when already imported above 1`] = `
"import Tabs from '@theme/Tabs';

import TabItem from '@theme/TabItem';

<Tabs>
  <TabItem value="npm">
    \`\`\`bash
      $ npm install --global docusaurus
    \`\`\`
  </TabItem>

  <TabItem value="yarn" label="Yarn">
    \`\`\`bash
      $ yarn global add docusaurus
    \`\`\`
  </TabItem>

  <TabItem value="pnpm" label="pnpm">
    \`\`\`bash
      $ pnpm add --global docusaurus
    \`\`\`
  </TabItem>
</Tabs>
"
`;

exports[`npm2yarn plugin does not re-import tabs components when already imported below 1`] = `
"<Tabs>
  <TabItem value="npm">
    \`\`\`bash
      $ npm install --global docusaurus
    \`\`\`
  </TabItem>

  <TabItem value="yarn" label="Yarn">
    \`\`\`bash
      $ yarn global add docusaurus
    \`\`\`
  </TabItem>

  <TabItem value="pnpm" label="pnpm">
    \`\`\`bash
      $ pnpm add --global docusaurus
    \`\`\`
  </TabItem>
</Tabs>

import Tabs from '@theme/Tabs';

import TabItem from '@theme/TabItem';
"
`;

exports[`npm2yarn plugin does not work when language is not set 1`] = `
"\`\`\`npm2yarn
npm install --save docusaurus-plugin-name
\`\`\`

\`\`\`bash
npm install --save docusaurus-plugin-name
\`\`\`

\`\`\`shell
npm install --save docusaurus-plugin-name
\`\`\`
"
`;

exports[`npm2yarn plugin work with custom converter 1`] = `
"import Tabs from '@theme/Tabs'
import TabItem from '@theme/TabItem'

## Installing a plugin

A plugin is usually a npm package, so you install them like other npm packages using npm.

<Tabs>
  <TabItem value="npm">
    \`\`\`bash
    npm install --save docusaurus-plugin-name
    \`\`\`
  </TabItem>

  <TabItem value="Turbo">
    \`\`\`bash
    turbo install --save docusaurus-plugin-name
    \`\`\`
  </TabItem>
</Tabs>
"
`;

exports[`npm2yarn plugin work with pnpm converter 1`] = `
"import Tabs from '@theme/Tabs'
import TabItem from '@theme/TabItem'

## Installing a plugin

A plugin is usually a npm package, so you install them like other npm packages using npm.

<Tabs>
  <TabItem value="npm">
    \`\`\`bash
    npm install --save docusaurus-plugin-name
    \`\`\`
  </TabItem>

  <TabItem value="pnpm" label="pnpm">
    \`\`\`bash
    pnpm add docusaurus-plugin-name
    \`\`\`
  </TabItem>
</Tabs>
"
`;

exports[`npm2yarn plugin work with yarn converter 1`] = `
"import Tabs from '@theme/Tabs'
import TabItem from '@theme/TabItem'

## Installing a plugin

A plugin is usually a npm package, so you install them like other npm packages using npm.

<Tabs>
  <TabItem value="npm">
    \`\`\`bash
    npm install --save docusaurus-plugin-name
    \`\`\`
  </TabItem>

  <TabItem value="yarn" label="Yarn">
    \`\`\`bash
    yarn add docusaurus-plugin-name
    \`\`\`
  </TabItem>
</Tabs>
"
`;

exports[`npm2yarn plugin works on installation file 1`] = `
"import Tabs from '@theme/Tabs'
import TabItem from '@theme/TabItem'

<Tabs>
  <TabItem value="npm">
    \`\`\`bash
      $ npm install --global docusaurus
    \`\`\`
  </TabItem>

  <TabItem value="yarn" label="Yarn">
    \`\`\`bash
      $ yarn global add docusaurus
    \`\`\`
  </TabItem>

  <TabItem value="pnpm" label="pnpm">
    \`\`\`bash
      $ pnpm add --global docusaurus
    \`\`\`
  </TabItem>
</Tabs>
"
`;

exports[`npm2yarn plugin works on plugin file 1`] = `
"import Tabs from '@theme/Tabs'
import TabItem from '@theme/TabItem'

## Installing a plugin

A plugin is usually a npm package, so you install them like other npm packages using npm.

<Tabs>
  <TabItem value="npm">
    \`\`\`bash
    npm install --save docusaurus-plugin-name
    \`\`\`
  </TabItem>

  <TabItem value="yarn" label="Yarn">
    \`\`\`bash
    yarn add docusaurus-plugin-name
    \`\`\`
  </TabItem>

  <TabItem value="pnpm" label="pnpm">
    \`\`\`bash
    pnpm add docusaurus-plugin-name
    \`\`\`
  </TabItem>
</Tabs>
"
`;

exports[`npm2yarn plugin works with common commands 1`] = `
"import Tabs from '@theme/Tabs'
import TabItem from '@theme/TabItem'

<Tabs groupId="npm2yarn">
  <TabItem value="npm">
    \`\`\`bash
    npm run xxx -- --arg
    \`\`\`
  </TabItem>

  <TabItem value="yarn" label="Yarn">
    \`\`\`bash
    yarn xxx --arg
    \`\`\`
  </TabItem>

  <TabItem value="pnpm" label="pnpm">
    \`\`\`bash
    pnpm run xxx --arg
    \`\`\`
  </TabItem>
</Tabs>

<Tabs groupId="npm2yarn">
  <TabItem value="npm">
    \`\`\`bash
    npm install package
    \`\`\`
  </TabItem>

  <TabItem value="yarn" label="Yarn">
    \`\`\`bash
    yarn add package
    \`\`\`
  </TabItem>

  <TabItem value="pnpm" label="pnpm">
    \`\`\`bash
    pnpm add package
    \`\`\`
  </TabItem>
</Tabs>

<Tabs groupId="npm2yarn">
  <TabItem value="npm">
    \`\`\`bash
    npm remove package-name
    \`\`\`
  </TabItem>

  <TabItem value="yarn" label="Yarn">
    \`\`\`bash
    yarn remove package-name
    \`\`\`
  </TabItem>

  <TabItem value="pnpm" label="pnpm">
    \`\`\`bash
    pnpm remove package-name
    \`\`\`
  </TabItem>
</Tabs>

<Tabs groupId="npm2yarn">
  <TabItem value="npm">
    \`\`\`bash
    npm init docusaurus
    npm init docusaurus@latest my-website classic
    \`\`\`
  </TabItem>

  <TabItem value="yarn" label="Yarn">
    \`\`\`bash
    yarn create docusaurus
    yarn create docusaurus@latest my-website classic
    \`\`\`
  </TabItem>

  <TabItem value="pnpm" label="pnpm">
    \`\`\`bash
    pnpm create docusaurus
    pnpm create docusaurus@latest my-website classic
    \`\`\`
  </TabItem>
</Tabs>
"
`;

exports[`npm2yarn plugin works with simplest md 1`] = `
"import Tabs from '@theme/Tabs'
import TabItem from '@theme/TabItem'

# Title

Hey

<Tabs>
  <TabItem value="npm">
    \`\`\`bash
    npm install test
    \`\`\`
  </TabItem>

  <TabItem value="yarn" label="Yarn">
    \`\`\`bash
    yarn add test
    \`\`\`
  </TabItem>

  <TabItem value="pnpm" label="pnpm">
    \`\`\`bash
    pnpm add test
    \`\`\`
  </TabItem>
</Tabs>
"
`;

exports[`npm2yarn plugin works with sync option 1`] = `
"import Tabs from '@theme/Tabs'
import TabItem from '@theme/TabItem'

## Installing a plugin

A plugin is usually a npm package, so you install them like other npm packages using npm.

<Tabs groupId="npm2yarn">
  <TabItem value="npm">
    \`\`\`bash
    npm install --save docusaurus-plugin-name
    \`\`\`
  </TabItem>

  <TabItem value="yarn" label="Yarn">
    \`\`\`bash
    yarn add docusaurus-plugin-name
    \`\`\`
  </TabItem>

  <TabItem value="pnpm" label="pnpm">
    \`\`\`bash
    pnpm add docusaurus-plugin-name
    \`\`\`
  </TabItem>
</Tabs>
"
`;
